From b4d17a6a3cf4ef572818fddd52269c83fb03f63a Mon Sep 17 00:00:00 2001
Message-ID: <b4d17a6a3cf4ef572818fddd52269c83fb03f63a.1715982305.git.amy@amyspark.me>
From: "L. E. Segovia" <amy@amyspark.me>
Date: Fri, 17 May 2024 21:39:10 +0000
Subject: [PATCH 2/2] dumper: Fix introspection binaries missing rpaths for
 relative dylibs

Apple expects dylibs to have an ID of the form `@rpath/libyadda.dylib`.
For this convention to work, a RPATH entry pointing to the library
prefix and/or to the relative location of said prefix w.r.t the
dependent executable or library, must be inserted.

This need previously did not arose because gobject-introspection
implicitly relied on the consumers having IDs set to their absolute
paths. This is no longer the case with GStreamer's Cerbero, which now
sanitizes RPATHs, load commands, and dylib IDs at install time.

The simplest and most obvious fix is to ensure that, for each library
load path, the corresponding rpath entry is inserted.
---
 giscanner/dumper.py | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/giscanner/dumper.py b/giscanner/dumper.py
index 74a494b..cdb5400 100644
--- a/giscanner/dumper.py
+++ b/giscanner/dumper.py
@@ -252,6 +252,12 @@ class DumpCompiler(object):
             args.extend(pkg_config_libs)
             self._compiler.get_external_link_flags(args, self._options.libraries)
 
+        if sys.platform == 'darwin':
+            # If the libraries' ID are of the form (@rpath/libfoo.dylib),
+            # then nothing previously can have added the needed rpaths
+            rpath_entries_to_add = [lib.replace('-L/', '-Wl,-rpath,/') for lib in pkg_config_libs if lib.startswith('-L/')]
+            args.extend(rpath_entries_to_add)
+
         if not self._compiler.check_is_msvc():
             for ldflag in shlex.split(os.environ.get('LDFLAGS', '')):
                 args.append(ldflag)
-- 
2.44.0

